tools/clang/lib/SPIRV/RemoveBufferBlockVisitor.cpp

```diff
@@ -71,15 +72,88 @@ bool RemoveBufferBlockVisitor::visitInstruction(SpirvInstruction *inst) {
 
   // For all instructions, if the result type is a pointer pointing to a struct
   // with StorageBuffer interface, the storage class must be updated.
-  if (auto *ptrResultType = dyn_cast<SpirvPointerType>(inst->getResultType())) {
-    if (hasStorageBufferInterfaceType(ptrResultType->getPointeeType()) &&
-        ptrResultType->getStorageClass() != spv::StorageClass::StorageBuffer) {
-      inst->setStorageClass(spv::StorageClass::StorageBuffer);
-      inst->setResultType(context.getPointerType(
-          ptrResultType->getPointeeType(), spv::StorageClass::StorageBuffer));
+  const auto *instType = inst->getResultType();
+  const auto *newInstType = instType;
+  spv::StorageClass newInstStorageClass = spv::StorageClass::Max;
+  if (updateStorageClass(instType, &newInstType, &newInstStorageClass)) {
+    inst->setResultType(newInstType);
+    inst->setStorageClass(newInstStorageClass);
+  }
+
+  return true;
+}
+
+bool RemoveBufferBlockVisitor::updateStorageClass(
+    const SpirvType *type, const SpirvType **newType,
+    spv::StorageClass *newStorageClass) {
+  if (!isa<SpirvPointerType>(type))
+    return false;
+
+  auto *ptrType = cast<SpirvPointerType>(type);
+  const auto *innerType = ptrType->getPointeeType();
+
+  // For usual cases such as _ptr_Uniform_StructuredBuffer_float.
+  if (hasStorageBufferInterfaceType(innerType) &&
+      ptrType->getStorageClass() != spv::StorageClass::StorageBuffer) {
+    *newType =
+        context.getPointerType(innerType, spv::StorageClass::StorageBuffer);
+    *newStorageClass = spv::StorageClass::StorageBuffer;
+    return true;
+  }
+
+  // For pointer-to-pointer cases (which need legalization), we could have a
+  // type like: _ptr_Function__ptr_Uniform_type_StructuredBuffer_float.
+  // In such cases, we need to update the storage class for the inner pointer.
```

Comment:
This code handles pointer-to-pointer cases e.g., `_ptr_Function__ptr_Uniform_type_StructuredBuffer_float`.
What happens if it is deeper pointer-to-pointer-to-pointer-to... ?

Do we need a recursive function or a for-loop for deeper cases that we do not even statically know the depth?

Reply:
In SPIR-V it's illegal to have pointer to pointers. We deliberately use the ptr-to-ptr scheme for cases that require spirv legalization (usually in the form of inlining). spirv-opt recognizes this pattern and performs legalization. So there should be no cases of more than 2 ptrs.

If for some reason there's a need for more than 2 levels in the future, then that would require a very large effort fixing many places in the codebase, including both dxc and spirv-opt.